homework_3/events/index.js

```diff
@@ -0,0 +1,97 @@
+const stream = require('stream');
+const fs = require('fs');
+const path = require('path');
+const csv = require('csv-parser');
+
+// --- --- --- --- --- --- --- Handlers --- --- --- --- --- --- --- ---
+
+getAll = async(req, res) => readEvents().then(events => {
+    const streamEvents = stream.Readable.from(JSON.stringify(events));
+    streamEvents.pipe(res);
+});
+
+getEvents = async(req, res) => {
+    const location = req.query.location;
+    readEvents().then(events => {
+        if (location)
+            events = events.filter(event => event.location === location);
+        res.json(events);
+    });
+}
+
+getEvent = async(req, res) => {
+    const id = Number(req.params.id);
+    if (Number.isInteger(id)) {
+        readEvents().then(events => res.json(events[id]));
+    } else {
+        res.json({ error: 'invalid param' });
+    }
+}
+
+remove = async(req, res) => {
+    const id = Number(req.params.id);
+    if (Number.isInteger(id)) {
+        readEvents().then(events => {
+            events = events.filter((e) => Number(e.id) !== id)
+            updateFile(events);
+            res.json({ success: true });
+        });
+    } else {
+        res.json({error: 'Invalid ID param'});
+    }
+}
+
+insert = async(req, res) => {
+    const { title, location, date, hour } = req.body;
+    readEvents().then(events => {
+        // Generate unique ID
+        events.push({id : events.length + 1, title, location, date, hour}); // update memory
+        updateFile(events); // update file
+        res.json([{ result: 'Event was added' }])
+    });
+}
+
+update = async(req, res) => {
+    const { title, location, date, hour } = req.body;
+    const id = Number(req.params.id);
+
+    readEvents().then(events => {
+        let event = events.find((e) => Number(e.id) === id)
+        if (event) {
+            event = { id, title, location, date, hour };
+            events = events.filter((e) => Number(e.id) !== id)
+            events.push(event);
+            updateFile(events);
+            res.json([{ result: 'Event was updated' }])
+        } else {
+            res.json({ error: 'Event not found' });
+        }
+    });
+}
+
+const sourcePath = path.join(__dirname, '..', 'data', 'events.csv');
+
+readEvents = () => {
```

Comment:
We could move methods that work with csv data to another module/file. We would divide responsibilities between modules/files. One file would be responsible for the HTTP layer and another one for data/csv layer.